### PR TITLE
Registration and Authentication

### LINKED ISSUE #10: Write cryptographic flow for authenticating a user
This should capture the expected communications between a local client and the key servers.

### LINKED ISSUE #11: Write cryptographic flow for registering a user
This should capture the expected communications between a local client and the key servers and detail the OPAQUE suite used.

### PR DESCRIPTION
**Pull request by indomitableSwan**
This is currently a draft of text to be included for registration and authentication. In terms of eventual placement, it may be good to include the registration.md text in the same document that covers the other operations the client can take (apart from authenticate, as this is handled separately).

There are many TODOS that need to be filled in OR linked to new issues. Mostly the former, I think.

This draft 
- take a very conservative approach to session handling: one request per session. This is obviously not an ideal choice for the final system, but I think for the PoC this is a reasonable (and temporary) default.
- does not attempt to write out the individual OPAQUE protocol flows.

Closes #11.
Closes #10.

### CODE REVIEW

File: systems-architecture.md [hunk 1]
```diff
@@ -32,23 +32,42 @@ We wish to ensure the following in our implementation:
...
+    1. The client receives as output two values, an `export_key` (matching that from the registration phase) and a `session_key` (which is the output of an authenticated key exchange).
+    1. The server receives as output a `session_key` matching that of the client.
+    1. Both client and server MUST receive confirmation of the success of the AKE.
+1. The client and server open an encrypted channel secured under a key derived from `session_key`. [TODO](https://github.com/boltlabs-inc/key-mgmt-spec/issues/29): Include additional details here.
+1. At this point, the registration session is considered _open_. 
+1. We have the following requirements for using an open registration session:
+    1. All messages sent between the client and server MUST be over this encrypted channel, i.e., all messages should be encrypted under the given shared key. 
+    1. The key server MUST additionally reject all messages sent over this channel that fail the AEAD verification checks.
+    1. The only valid request for a registration is a request to [complete registration](TODO: fill me in).
```
Comment (marsella) [inline systems-architecture.md:55]: What does this mean? I thought at this point we could go on to either terminate the session or make a request (store, retrieve, audit, import, export) as below. I thought registration was already complete at this stage.

File: systems-architecture.md [hunk 2]
```diff
@@ -32,23 +32,42 @@ We wish to ensure the following in our implementation:
...
+    1. The key server MUST close the session after completion of the complete registration request.
+        - [TODO](https://github.com/boltlabs-inc/key-mgmt-spec/issues/51): allow more than one request. See similar TODO note below.
```
Comment (marsella) [inline systems-architecture.md:57]: Added #51.

File: systems-architecture.md (hunk 1 above)
Comment (indomitableSwan) [reply systems-architecture.md:55]: So to complete registration we have to generated the storage_key and store it at the server. I may have been inconsistent in my draft on how to write about this.

Comment (indomitableSwan) [reply systems-architecture.md:55]: L18 in register.md is talking about what "complete registration" requests are.

Comment (marsella) [reply systems-architecture.md:55]: Oh, I see. Fixed the link.

File: register.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    1. Generates `user_id`, a globally unique identifier (GUID). 
+        - [TODO](https://github.com/boltlabs-inc/key-mgmt-spec/issues/52): Specify how user IDs are created.
+    1. [Opens a registration session](systems-architecture.md#opening-a-registration-session) with the key server.
+    1. Derives  a symmetric `opaque_encryption_key` for [`Enc`](#external-dependencies) from the `export_key`.
+    The opaque encryption key will have length `len` bytes, the default for `Enc`.
+        1. Set `context = "opaque-derived stored client key for <AEAD choice and params>"`.
+        1. Compute `secret = HKDF("opaque encryption key", export_key, context, len)`.
+        1. Define `opaque_encryption_key = (secret, context)`.
+    1. Runs the [generate protocol](cryptographic_flows.md#generate-a-secret) to get a symmetric key `storage_key` for [`Enc`](#external-dependencies) of length 32 bytes.
+        - The `storage key` MUST NOT be saved, stored, or used in any context outside this protocol. It must not be passed to the calling application.
```
Comment (marsella) [inline register.md:14]: Is this a thorough enough warning?

**Review by marsella — APPROVED** [review]

**Pull request merged**